src/Mvc/Mvc.RazorPages/src/Infrastructure/PageLoaderMatcherPolicy.cs

```diff
@@ -73,17 +72,44 @@ public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context,
             for (var i = 0; i < candidates.Count; i++)
             {
                 ref var candidate = ref candidates[i];
-                var endpoint = (RouteEndpoint)candidate.Endpoint;
+                var endpoint = candidate.Endpoint;
 
                 var page = endpoint.Metadata.GetMetadata<PageActionDescriptor>();
                 if (page != null)
                 {
-                    var compiled = _loader.Load(page);
-                    candidates.ReplaceEndpoint(i, compiled.Endpoint, candidate.Values);
+                    // We found an endpoint instance that has a PageActionDescriptor, but not a
+                    // CompiledPageActionDescriptor. Update the CandidateSet.
+                    var compiled = _loader.LoadAsync(page);
+                    if (compiled.IsCompletedSuccessfully)
+                    {
+                        candidates.ReplaceEndpoint(i, compiled.Result.Endpoint, candidate.Values);
+                    }
+                    else
+                    {
+                        // In the most common case, LoadAsync will return a synchronous result.
+                        // Avoid going async since this is a fairly hot path.
+                        return ApplyAsyncAwaited(candidates);
```

Comment:
In ApplyAsyncAwaited you're recalling LoadAsync for event candidate again. Shouldn't you pass in the last task and the index so you don't make unnecessary calls to LoadAsync? Or do you think it is not worth the complexity?

Reply:
Ended up doing the continue. I was worried it might have appeared to be more complex than necessary, but it's fairly trivial code .